holoviews/util/__init__.py

```diff
@@ -105,6 +105,92 @@ def __call__(self, *args, **params):
             return StoreOptions.set_options(obj, options)
 
 
+    @classmethod
+    def expand_options(cls, options, backend=None):
+        """
+        Validates and expands a dictionaries of options indexed by
+        type[.group][.label] keys into separate style, plot and norm
+        options.
+
+            opts.expand_options({'Image': dict(cmap='viridis', show_title=False)})
+
+        returns
+
+            {'Image': {'plot': dict(show_title=False), 'style': dict(cmap='viridis')}}
+        """
+        current_backend = Store.current_backend
+        backend_options = Store.options(backend=backend or current_backend)
+        expanded = {}
+        for objspec, options in options.items():
+            objtype = objspec.split('.')[0]
+            if objtype not in backend_options:
+                raise ValueError('%s type not found, could not apply options.'
+                                 % objtype)
+            obj_options = backend_options[objtype]
+            expanded[objspec] = {g: {} for g in obj_options.groups}
+            for opt, value in options.items():
+                found = False
+                valid_options = []
+                for g, group_opts in obj_options.groups.items():
+                    if opt in group_opts.allowed_keywords:
+                        expanded[objspec][g][opt] = value
+                        found = True
+                        break
+                    valid_options += group_opts.allowed_keywords
+                if found: continue
+                cls._options_error(opt, objtype, backend, valid_options)
+        return expanded
+
+
+    @classmethod
+    def _options_error(cls, opt, objtype, backend, valid_options):
+        """
+        Generates an error message for an invalid option suggesting
+        similar options through fuzzy matching.
+        """
```

Comment:
I should have raised this earlier. Isn't all this logic what [``OptionError``](https://github.com/ioam/holoviews/blob/master/holoviews/core/options.py#L60) is for?  The logic is certainly closely related so maybe ``OptionError`` should be enhanced or extended somehow...

Reply:
I did look into that but I can be much more eager about the validation here because the user can declare an explicit backend and because it has to resolve each option into the appropriate group. I think we should look at unifying the validation, especially once the magic allows declaring an explicit backend.

What I would like to know is why the OptionError handling is so horrendously complex. Why does it have to record errors? It seems like that code would be hugely simplified by a utility that checks whether an option is invalid across all backends, so you can eagerly raise as soon as you encounter an unsupported option rather than processing all the options and checking validity at the end.

Separately I'm still confused in what cases the existing code actually validates anything, it seems like ``.opts`` still has zero validation, since this raises no error:

``hv.Curve([]).opts(plot=dict(some_nonexistent_option=3))``

Reply:
> What I would like to know is why the OptionError handling is so horrendously complex. Why does it have to record errors? It seems like that code would be hugely simplified by a utility that checks whether an option is invalid across all backends, so you can eagerly raise as soon as you encounter an unsupported option rather than processing all the options and checking validity at the end.

The main reason it is this way is because the errors should be derived from the ``allowed_keywords`` specified in the ``Options``. It is true that these allowed keywords are set from the plot and style option declarations which is what your proposed utility would rely on directly. I'm certainly not opposed to simplifying the code with a more direct check but the current design allows more flexibility by having a level of indirection: plot and style option declarations -> allowed keywords on options and errors-> validation.

As I agree we rarely use the extra flexibility that this design affords us, I would be happy to see a PR that simplifies things to  plot and style option declarations -> validation. My worry is that such a PR might have to touch a lot more than you might expect (e.g how the errors are processed and printed by the %opts magic(s)).

Reply:
Right, I definitely want to tackle and simplify that code because I'm having a really tough time following it and I think I'll much prefer an approach that eagerly errors when it hits an unsupported option.

Currently I'm leaning toward merging this PR and getting a dev release out and then to follow up with a PR to try to unify the option validation. What do you think?

Reply:
> I'll much prefer an approach that eagerly errors when it hits an unsupported option.

That is exactly how it worked when we only ever had to worry about a single backend at a time. The complexity was introduced when we decided to merge the allowed keywords across backends which was implemented in a way to extend the old system (e.g where the magics used ``OptionError``  message to pretty print suggestions).

> Currently I'm leaning toward merging this PR and getting a dev release out and then to follow up with a PR to try to unify the option validation. What do you think?

Makes sense. I don't think of it as 'unifying' the option validation though (it is already unified, which is what makes it ugly). That PR would be more of a refactor/redesign by validating without looking at the allowed keywords for ``Options``.

I suppose one thing the current design lets you do is to declare options for elements/plots that don't exist, using ``allowed_keywords`` to do validation. The current design keeps the option system decoupled from elements/plots with ``allow_keywords`` acting as the link between them (set somewhere on ``Store`` iirc). This (in theory) could let you use the option system in a completely different project/library.

In practice, this decoupling doesn't seem to do much useful (unless you can imagine using the options system for something without needing a corresponding element/plot). In other words, I think there is a sound design choice behind the way it is right now but 1) it doesn't add much of value in practice 2) it is becomes complicated once you start taking the union between backends. For this reason, I would be happy to see things simplified even if it means coupling the options system closer to the elements and plot classes.

Reply:
Don't think we necessarily need to couple the two closer together (although I haven't considered all the implications). I just feel like it comes down to the way you iterate over the backends. Currently it seems like looping over the backends is the primary loop, which means you always iterate over all backends and can't raise until you've processed them all. The approach I've taken is to validate the current (or explicitly selected) backend first, and if an option is invalid for that backend it falls back to checking whether the option is valid for other backends. This means that as soon as an invalid option is encountered it can raise an error (or warning) rather than having to process everything first and only asserting whether an option was invalid at the end.

My only question then is about the OptsMagic, is there any reason why that can't raise eagerly or does it somehow have to catch errors and reraise them in a special way?

Reply:
> does it somehow have to catch errors and reraise them in a special way?

That is exactly what it does.

Reply:
Okay, I can still do that easily enough.

Reply:
Though I will note it does it using the [``validation_error_message``](https://github.com/ioam/holoviews/blob/master/holoviews/core/options.py#L1420) method which might mean there is indeed one single place where your approach could fit in nicely. Of course, there may still be other consequences I haven't thought about...e.g how will normal error validation work if you work with option trees directly?

Reply:
When you work with an OptionTree directly the allowed_keywords are well defined and no cross-backend validation is needed so that should be fine. I'll open a PR branching off this one to see what it could look like.